ttnn/cpp/ttnn/operations/ccl/erisc_datamover_builder.cpp

```diff
@@ -0,0 +1,414 @@
+        auto& backward_direction_edm = edm_builders_backward_direction.at(device_sequence[i]->id());
+
+        for (size_t l = 0; l < num_links; l++) {
+            forward_direction_edm.at(l).connect_to_downstream_edm(backward_direction_edm.at(l));
+            backward_direction_edm.at(l).connect_to_downstream_edm(forward_direction_edm.at(l));
+        }
+    }
+
+}
+
+
+SenderWorkerAdapterSpec EdmLineFabricOpInterface::uniquely_connect_worker(Device* device, Direction direction) {
+    TT_ASSERT((direction == FORWARD) ? edm_builders_forward_direction.find(device->id()) != edm_builders_forward_direction.end()
+                                     : edm_builders_backward_direction.find(device->id()) != edm_builders_backward_direction.end());
+    auto& edm_builders = (direction == FORWARD) ? edm_builders_forward_direction.at(device->id())
+                                                : edm_builders_backward_direction.at(device->id());
+    auto &link_count_map = (direction == FORWARD) ? next_forward_direction_edm_available : next_backward_direction_edm_available;
+    const auto next_link = link_count_map[device->id()];
+    link_count_map[device->id()] = next_link + 1;
+
+    TT_ASSERT(edm_builders.size() > 0);
+    TT_ASSERT(next_link < edm_builders.size());
+    return edm_builders.at(next_link).build_connection_to_worker_channel();
+}
+
+void EdmLineFabricOpInterface::build_kernels() const {
+    TT_ASSERT(device_sequence.size() == programs.size());
+    for (size_t i = 0; i < device_sequence.size(); i++) {
+        Program* program = programs[i];
+        Device* device = device_sequence[i];
+        if (edm_builders_forward_direction.find(device_sequence[i]->id()) != edm_builders_forward_direction.end()) {
+            for (auto& edm_builder : edm_builders_forward_direction.at(device_sequence[i]->id())) {
+                auto local_edm_kernel = ttnn::ccl::generate_edm_kernel(
+                    *program,                             // sender_program,
+                    device,                              // sender_device,
+                    edm_builder,  // chip_0_edm_builder,
+                    edm_builder.my_eth_core_logical,         // eth_sender_core,
+                    NOC::NOC_0);
+            }
+        }
+        if (edm_builders_backward_direction.find(device_sequence[i]->id()) != edm_builders_backward_direction.end()) {
+            for (auto& edm_builder : edm_builders_backward_direction.at(device_sequence[i]->id())) {
+                auto local_edm_kernel = ttnn::ccl::generate_edm_kernel(
+                    *program,                             // sender_program,
+                    device,                              // sender_device,
+                    edm_builder,  // chip_0_edm_builder,
+                    edm_builder.my_eth_core_logical,         // eth_sender_core,
+                    NOC::NOC_0);
+            }
+        }
```

Comment:
the only difference between these two is in edm_builders_forward_direction vs edm_builders_backward_direction, right? consider deduplicating the code